chunkset.c

```diff
@@ -36,8 +44,12 @@ static inline void chunkmemset_4(uint8_t *from, chunk_t *chunk) {
 }
 
 static inline void chunkmemset_8(uint8_t *from, chunk_t *chunk) {
-#ifdef UNALIGNED_OK
+#if defined(UNALIGNED64_OK)
     *chunk = *(uint64_t *)from;
+#elif defined(UNALIGNED_OK)
+    uint64_t full_chunk = *(uint64_t *)from;
```

Comment:
This is still possibly 64-bit unaligned read.

Reply:
The fact that it's *declared* as an uint64_t on the stack *should* mean it's properly aligned, though, right?

Reply:
Oh, wait, the *read*, not the store. :D.

Reply:
The problem is ```from``` which is not guaranteed to be on stack and even if it is on stack, it might not be aligned on 64-bit boundary. Stack can be aligned on either 32-bit or 64-bit boundary depending on system.

On release builds, optimizer might not generate any code for this line, but debug builds must preserve all local variables that are used.

Reply:
Although, `from` itself is an ` uint8_t *`, so it's probably never going to be an issue?

(My brain hurts >_<).

In any case, GCC appears to be chunking that in two 32bits loads anyway:

```
3980   │ static inline void chunkmemset_8(uint8_t *from, chunk_t *chunk) {
3981   │ #if defined(UNALIGNED64_OK)
3982   │     *chunk = *(uint64_t *)from;
3983   │ #elif defined(UNALIGNED_OK)
3984   │     uint64_t full_chunk = *(uint64_t *)from;
3985   │     3bb2:   682c        ldr r4, [r5, #0]
3986   │     3bb4:   686b        ldr r3, [r5, #4]
```

Reply:
I guess I can make that explicit by doing it this way instead?

```C
    uint32_t* p = (uint32_t *)from;
    *chunk[0] = *p;
    *chunk[1] = *(p+1);
```

Reply:
Being ```uint8_t *``` only implies that the pointer is assumed to be unaligned. When you cast it to wider type, you essentially create temporary variable that must be assigned to another variable with same alignment restrictions.... Basically you can only assign the value of memory it references to, to vector of ```uint8_t``` or another memory location pointed by another variable of type ```uint8_t *```. Assigning values explicitly to registers was removed from recent gcc versions, even though compiler itself might temporarily use registers to store the variable contents.

Basically this code relies on compiler generating the correct code instead of writing it as two 32-bit copies explicitly.

Reply:
Or even...

```
uint32_t* p = (uint32_t *)from;
chunk[0] = p[0];
chunk[1] = p[1];
```

Reply:
This only has a minor register usage impact once inlined in `chunkcopy_c`:

```
4066   │     uint64_t full_chunk = *(uint64_t *)from;
4067   │     3c0e:   680c        ldr r4, [r1, #0]
4068   │     3c10:   684b        ldr r3, [r1, #4]
4069   │     ((uint32_t *)out)[0] = *chunk[0];
4070   │     3c12:   6004        str r4, [r0, #0]
4071   │     ((uint32_t *)out)[1] = *chunk[1];
4072   │     3c14:   6043        str r3, [r0, #4]
```

```
4067   │     *chunk[0] = *p;
4068   │     3c0e:   680c        ldr r4, [r1, #0]
4069   │     *chunk[1] = *(p+1);
4070   │     3c10:   684b        ldr r3, [r1, #4]
4071   │     ((uint32_t *)out)[0] = *chunk[0];
4072   │     3c12:   6004        str r4, [r0, #0]
4073   │     ((uint32_t *)out)[1] = *chunk[1];
4074   │     3c14:   6043        str r3, [r0, #4]
```

Reply:
And a minor one (for the better?) in `CHUNKMEMSET`:

```
6140   │     uint64_t full_chunk = *(uint64_t *)from;
6141   │     4af8:   682a        ldr r2, [r5, #0]
6142   │     4afa:   686b        ldr r3, [r5, #4]
6143   │     *chunk[0] = (uint32_t)(full_chunk & 0xffffffff);
6144   │     4afc:   60ba        str r2, [r7, #8]
6145   │     *chunk[1] = (uint32_t)(full_chunk >> 32);
6146   │     4afe:   613b        str r3, [r7, #16]
```

```
6143   │     *chunk[0] = *p;
6144   │     4af8:   682b        ldr r3, [r5, #0]
6145   │     4afa:   60bb        str r3, [r7, #8]
6146   │     *chunk[1] = *(p+1);
6147   │     4afc:   686b        ldr r3, [r5, #4]
6148   │     4afe:   613b        str r3, [r7, #16]
```

Reply:
@mtl1979: Gah, I missed your comments, but it looks like we had mostly the same idea at around the same time ;).

Reply:
Now that I'm caught up, is there a preference towards the array index notation vs. the pointer arithmetic one?

Reply:
When one of the variables is type[], then it makes more sense to use array notation on both sides even if the other variable is type* as it is more consistent and easier to read. Pointer arithmetics should only be used when all variables are type*, or when using operator++ etc. to advance the pointer itself.